src/ips.rs

```diff
@@ -0,0 +1,526 @@
+            }
+
+            // 2 - Calculate desired vertex degree
+            // In the first iteration we will use avg of degree and degree_delta so all
+            // nodes should pursue to degree_delta level. That could be bad if graph's vertexes
+            // have very high (or low) degrees and therefore, delta is very high (or low) too. But until
+            // we have some better idea this one is the best we can do to keep up with the graph.
+            let desired_degree = (degree_avg as u32 + degrees.get(&node_ip).unwrap()) / 2;
+
+            // 3 - Calculate how many peers to add or delete from peerlist
+            let peers_to_delete_count = if desired_degree < *degrees.get(&node_ip).unwrap() {
+                degrees.get(&node_ip).unwrap() - desired_degree
+            } else {
+                // Check if config forces to change peerlist even if we have good degree.
+                // This should be always set to at least one to allow for some changes in graph -
+                // searching for better potential peers.
+                self.config.change_at_least
+            };
+
+            // Calculating how many peers should be added. If we have more peers than desired degree
+            // we will add at least config.change_at_least peers.
+            let mut peers_to_add_count = if desired_degree > *degrees.get(&node_ip).unwrap() {
+                desired_degree - degrees.get(&node_ip).unwrap() + peers_to_delete_count
+            } else {
+                self.config.change_at_least
+            };
+
+            // Limit number of changes to config value
+            if peers_to_add_count > self.config.change_no_more {
+                peers_to_add_count = self.config.change_no_more;
+            }
+
+            // 4 - Choose peers to delete from peerlist (based on ranking)
+            if peers_to_delete_count > 0 {
+                // Sort peers by rating (highest first)
+                curr_peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+                // Remove peers with lowest rating
+                for _ in 0..peers_to_delete_count {
+                    let peer_to_delete = curr_peer_ratings.pop();
+                    if let Some(peer_to_delete) = peer_to_delete {
+                        peer_list_entry.list.retain(|x| x != &peer_to_delete.0);
+                    }
+                }
+            }
+
+            // 5 - Find peers to add from selected peers (based on rating)
+            if peers_to_add_count > 0 {
+                // Take twice as many candidates as we need to add to peerlist to be able to
+                // choose best ones from them.
+                let mut candidates_to_search = peers_to_add_count * 2;
+
+                let mut candidates: Vec<(IpAddr, usize, f64)> =
+                    Vec::with_capacity(candidates_to_search as usize);
+
+                // Sort peers by rating
+                peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
```

Comment:
since peer_ratings is copied for every node, we are performing the same sort multiple times

Also, it means we would choose only a handful of top nodes by rating and put those on each suggested peer list where some node 🕵️‍♀️

so if we have 1000 peers, and 500 peers need to add ~5 nodes to improve their average degree,
all 500 peer list suggestions will contain the same (the best) top ~10 (since 5*2) nodes from the total of 1000 nodes

Reply:
That's not the same sort for each node. 
For each node `peer_ratings` are created from `const_factors` and then, each peer rating is adjusted based on distance between current node and that peer.
That makes `peer_ratings` list unique for each node.

Reply:
right, so my previous comment is true only when the `location` factor is set to zero.

but let me try to think about this like this:
so the `location` factor is `0.1`

which means that the `const_factors` contain a list of ratings from `0 - 90.0` - and that list, once sorted, is pretty static.
then if we calculate location factors for all nodes separately and sort them, we get a list with a rating range: `0 - 10.0`

when we add those two vectors together, the overall node rating won't change that much, since the variable part (the location) only makes a 0-10 difference.

so that means the top 20 best nodes will still be pretty similar in the final `peer_ratings` list.
so I propose we increase the location_ration to a higher value to make sure it breaks the static sorted list of `const_factors` ratings. so the top 20 nodes are much more different from node to node.

Reply:
Default weights changed.